File: grm_management/grm_management/page/space_calendar/space_calendar.js
frappe.pages['space_calendar'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Space Calendar',
		single_column: true
	});

	new SpaceCalendar(page);
}

class SpaceCalendar {
	constructor(page) {
		this.page = page;
		this.calendar = null;
		this.setup_page();
		this.make_calendar();
	}

	setup_page() {
		// Add filter for location
		this.page.add_field({
			fieldname: 'location',
			label: __('Location'),
			fieldtype: 'Link',
			options: 'GRM Location',
			change: () => this.refresh_calendar()
		});

		// Add filter for space type
		this.page.add_field({
			fieldname: 'space_type',
			label: __('Space Type'),
			fieldtype: 'Link',
			options: 'GRM Space Type',
			change: () => this.refresh_calendar()
		});

		// Add filter for space
		this.page.add_field({
			fieldname: 'space',
			label: __('Space'),
			fieldtype: 'Link',
			options: 'GRM Space',
			change: () => this.refresh_calendar()
		});

		// Add view switcher
		this.page.add_field({
			fieldname: 'view_type',
			label: __('View'),
			fieldtype: 'Select',
			options: ['Month', 'Week', 'Day', 'List'],
			default: 'Week',
			change: () => this.change_view()
		});

		// Add refresh button
		this.page.set_primary_action(__('New Booking'), () => {
			frappe.new_doc('GRM Booking');
		});

		this.page.add_button(__('Refresh'), () => {
			this.refresh_calendar();
		});
	}

	make_calendar() {
		this.$calendar = $('<div class="grm-space-calendar">').appendTo(this.page.main);

		this.$calendar.fullCalendar({
			header: {
				left: 'prev,next today',
				center: 'title',
				right: 'month,agendaWeek,agendaDay,listWeek'
			},
			defaultView: 'agendaWeek',
			editable: false,
			selectable: true,
			selectHelper: true,
			allDaySlot: false,
			slotDuration: '00:30:00',
			minTime: '06:00:00',
			maxTime: '22:00:00',
			businessHours: {
				dow: [0, 1, 2, 3, 4, 5, 6],
				start: '08:00',
				end: '20:00'
			},
			eventLimit: true,
			events: (start, end, timezone, callback) => {
				this.get_events(start, end, callback);
			},
			select: (start, end) => {
				this.create_booking(start, end);
			},
			eventClick: (event) => {
				frappe.set_route('Form', 'GRM Booking', event.id);
			},
			eventRender: (event, element) => {
				element.find('.fc-title').html(event.title);
				element.attr('title', event.description);
			}
		});
	}

	get_events(start, end, callback) {
		let filters = {
			booking_date: ['between', [
				moment(start).format('YYYY-MM-DD'),
				moment(end).format('YYYY-MM-DD')
			]]
		};

		// Apply additional filters
		let location = this.page.get_field('location').get_value();
		let space_type = this.page.get_field('space_type').get_value();
		let space = this.page.get_field('space').get_value();

		if (space) {
			filters.space = space;
		}

		frappe.call({
			method: 'grm_management.grm_management.page.space_calendar.space_calendar.get_bookings',
			args: {
				start: moment(start).format('YYYY-MM-DD'),
				end: moment(end).format('YYYY-MM-DD'),
				location: location,
				space_type: space_type,
				space: space
			},
			callback: (r) => {
				if (r.message) {
					let events = r.message.map(booking => {
						return {
							id: booking.name,
							title: `${booking.space_name} - ${booking.member_name}`,
							start: `${booking.booking_date}T${booking.start_time}`,
							end: `${booking.booking_date}T${booking.end_time}`,
							color: this.get_status_color(booking.status),
							description: `
								<strong>${booking.space_name}</strong><br>
								Member: ${booking.member_name}<br>
								Status: ${booking.status}<br>
								Duration: ${booking.duration_hours} hours<br>
								Amount: ${format_currency(booking.total_amount)}
							`
						};
					});
					callback(events);
				}
			}
		});
	}

	get_status_color(status) {
		const colors = {
			'Draft': '#9e9e9e',
			'Confirmed': '#2196f3',
			'Checked-in': '#4caf50',
			'Checked-out': '#607d8b',
			'No-show': '#ff9800',
			'Cancelled': '#f44336'
		};
		return colors[status] || '#9e9e9e';
	}

	create_booking(start, end) {
		let space = this.page.get_field('space').get_value();

		let booking = frappe.model.get_new_doc('GRM Booking');
		booking.booking_date = moment(start).format('YYYY-MM-DD');
		booking.start_time = moment(start).format('HH:mm:ss');
		booking.end_time = moment(end).format('HH:mm:ss');

		if (space) {
			booking.space = space;
		}

		frappe.set_route('Form', 'GRM Booking', booking.name);
	}

	refresh_calendar() {
		this.$calendar.fullCalendar('refetchEvents');
	}

	change_view() {
		let view_type = this.page.get_field('view_type').get_value();
		let view_map = {
			'Month': 'month',
			'Week': 'agendaWeek',
			'Day': 'agendaDay',
			'List': 'listWeek'
		};
		this.$calendar.fullCalendar('changeView', view_map[view_type]);
	}
}

File: grm_management/grm_management/page/space_calendar/space_calendar.py
# -*- coding: utf-8 -*-
# Copyright (c) 2026, Frappe Technologies and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _

@frappe.whitelist()
def get_bookings(start, end, location=None, space_type=None, space=None):
	"""Get bookings for calendar view"""
	filters = {
		'booking_date': ['between', [start, end]]
	}

	if space:
		filters['space'] = space
	elif space_type or location:
		# Get spaces matching criteria
		space_filters = {}
		if location:
			space_filters['location'] = location
		if space_type:
			space_filters['space_type'] = space_type

		spaces = frappe.get_all('GRM Space', filters=space_filters, pluck='name')
		if spaces:
			filters['space'] = ['in', spaces]
		else:
			return []

	bookings = frappe.get_all('GRM Booking',
		filters=filters,
		fields=[
			'name', 'booking_date', 'start_time', 'end_time',
			'space', 'member', 'status', 'duration_hours', 'total_amount'
		],
		order_by='booking_date, start_time'
	)

	# Enrich with space and member names
	for booking in bookings:
		booking['space_name'] = frappe.db.get_value('GRM Space', booking['space'], 'space_name')
		booking['member_name'] = frappe.db.get_value('GRM Member', booking['member'], 'full_name')

	return bookings

@frappe.whitelist()
def get_space_availability(date, location=None, space_type=None):
	"""Get space availability for a specific date"""
	# Get all spaces
	filters = {}
	if location:
		filters['location'] = location
	if space_type:
		filters['space_type'] = space_type

	filters['status'] = ['in', ['Available', 'Rented']]
	filters['allow_booking'] = 1

	spaces = frappe.get_all('GRM Space',
		filters=filters,
		fields=['name', 'space_name', 'space_type', 'capacity', 'status', 'hourly_rate']
	)

	# Get bookings for this date
	bookings = frappe.get_all('GRM Booking',
		filters={
			'booking_date': date,
			'status': ['not in', ['Cancelled', 'No-show']]
		},
		fields=['space', 'start_time', 'end_time', 'status']
	)

	# Create booking map
	booking_map = {}
	for booking in bookings:
		if booking['space'] not in booking_map:
			booking_map[booking['space']] = []
		booking_map[booking['space']].append({
			'start': booking['start_time'],
			'end': booking['end_time'],
			'status': booking['status']
		})

	# Add availability info to spaces
	for space in spaces:
		space['bookings'] = booking_map.get(space['name'], [])
		space['is_available'] = len(space['bookings']) == 0

	return spaces

@frappe.whitelist()
def check_space_conflict(space, booking_date, start_time, end_time, exclude_booking=None):
	"""Check if there's a booking conflict for a space"""
	filters = {
		'space': space,
		'booking_date': booking_date,
		'status': ['not in', ['Cancelled', 'No-show']]
	}

	if exclude_booking:
		filters['name'] = ['!=', exclude_booking]

	existing_bookings = frappe.get_all('GRM Booking',
		filters=filters,
		fields=['start_time', 'end_time', 'name']
	)

	for booking in existing_bookings:
		# Check for time overlap
		if (start_time < booking['end_time'] and end_time > booking['start_time']):
			return {
				'conflict': True,
				'booking': booking['name'],
				'message': _('This space is already booked from {0} to {1}').format(
					booking['start_time'], booking['end_time']
				)
			}

	return {'conflict': False}
